File: core/src/main/java/com/oneandone/typedrest/Endpoint.java
package com.oneandone.typedrest;

import com.damnhandy.uri.template.UriTemplate;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.*;
import java.util.Map;
import java.util.Set;
import org.apache.http.client.fluent.*;

/**
 * REST endpoint, i.e. a remote HTTP resource.
 */
public interface Endpoint {

    /**
     * The HTTP URI of the remote resource.
     *
     * @return The HTTP URI of the remote resource.
     */
    URI getUri();

    /**
     * The REST executor used to communicate with the remote resource.
     *
     * @return The REST executor used to communicate with the remote resource.
     */
    Executor getExecutor();

    /**
     * Controls the serialization of entities sent to and received from the server.
     *
     * @return Controls the serialization of entities sent to and received from the server.
     */
    ObjectMapper getSerializer();

    /**
     * Retrieves all links with a specific relation type cached from the last
     * request.
     *
     * @param rel The relation type of the links to look for.
     * @return The hrefs of the links resolved relative to this endpoint's URI.
     */
    Set<URI> getLinks(String rel);

    /**
     * Retrieves all links (with titles) with a specific relation type cached
     * from the last request.
     *
     * @param rel The relation type of the links to look for.
     * @return A map of hrefs (resolved relative to this endpoint's URI) to
     * titles (may be <code>null</code>).
     */
    Map<URI, String> getLinksWithTitles(String rel);

    /**
     * Retrieves a single link with a specific relation type.
     *
     * Uses cached data from last response if possible. Tries lazy lookup with
     * HTTP HEAD on cache miss.
     *
     * @param rel The relation type of the link to look for.
     * @return The href of the link resolved relative to this endpoint's URI.
     * @throws RuntimeException No link with the specified relation type could
     * be found.
     */
    URI link(String rel);

    /**
     * Retrieves a link template with a specific relation type.
     *
     * Uses cached data from last response if possible. Tries lazy lookup with
     * HTTP HEAD on cache miss.
     *
     * @param rel The relation type of the template to look for.
     * @return The unresolved link template.
     * @throws RuntimeException No link template with the specified relation
     * type could be found.
     */
    UriTemplate linkTemplate(String rel);

    /**
     * Retrieves a link template with a specific relation type and resolves it.
     *
     * @param rel The relation type of the template to look for.
     * @param variableName The name of the variable to insert.
     * @param value The value to insert for the variable.
     * @return The href of the link resolved relative to this endpoint's URI.
     * @throws RuntimeException No link template with the specified relation
     * type could be found.
     */
    default URI linkTemplate(String rel, String variableName, Object value) {
        return getUri().resolve(linkTemplate(rel).set(variableName, value).expand());
    }
}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/Endpoint.kt
package net.typedrest.endpoints

import net.typedrest.errors.ErrorHandler
import net.typedrest.errors.NotFoundException
import net.typedrest.links.LinkExtractor
import net.typedrest.serializers.Serializer
import okhttp3.OkHttpClient
import java.net.URI
import kotlin.collections.List

/**
 * Represents an endpoint, i.e., a remote HTTP resource.
 */
interface Endpoint {
    /**
     * The HTTP URI of the remote resource.
     */
    val uri: URI

    /**
     * The HTTP client used to communicate with the remote resource.
     */
    val httpClient: OkHttpClient

    /**
     * A list of serializers used for entities received from the server,
     * sorted from most to least preferred. Always uses first for sending to the server.
     */
    val serializers: List<Serializer>

    /**
     * Handles errors in responses.
     */
    val errorHandler: ErrorHandler

    /**
     * Extracts links from responses.
     */
    val linkExtractor: LinkExtractor

    /**
     * Resolves all links with a specific relation type. Uses cached data from last response.
     *
     * @param rel The relation type of the links to look for.
     * @return A list of pairs of URI and optional title.
     */
    fun getLinks(rel: String): List<Pair<URI, String?>>

    /**
     * Resolves a single link with a specific relation type. Uses cached data from last response if possible.
     *
     * Tries lazy lookup with HTTP HEAD on cache miss.
     *
     * @param rel The relation type of the link to look for.
     * @throws NotFoundException if no link with the specified relation could be found.
     * @return The URI of the link.
     */
    @Throws(NotFoundException::class)
    fun link(rel: String): URI

    /**
     * Resolves a link template with a specific relation type. Uses cached data from last response if possible.
     *
     * Tries lazy lookup with HTTP HEAD on cache miss.
     *
     * @param rel The relation type of the link template to look for.
     * @param variables Variables for resolving the template.
     * @throws NotFoundException if no link template with the specified relation could be found.
     * @return The href of the link resolved relative to this endpoint's URI.
     */
    @Throws(NotFoundException::class)
    fun linkTemplate(rel: String, variables: Map<String, Any>): URI
}
